### ISSUE TITLE
List of Modules should show required contrib modules

### ISSUE THREAD

**Issue by yorkshire-pudding** [backdrop/backdrop-issues#6173]
## Description of the need
I came across the issue recently where I couldn't find an installed module in the module page.  Bee showed it as enabled and other modules showed that they were required by this module.

The module in question is a Feeds plugin and the `feeds_system_info_alter()` function makes plugin modules for which an importer exists (even if the importer is disabled) a required module.  I only found this out because I dug into the info blob in the database and then found the required flag and the explanation.

It feels like Core UI should display contrib required modules in the module list but prevent disabling in the same way as when a module is a dependency.  Perhaps there could be another line with the dependency listing that uses the explanation that Feeds has given it?

I imagine as there is a hook this could be used by other module to add the required flag.

## Proposed solution

Required contrib modules are listed
They are prevented from being disabled and the explanation is given

## Alternatives that have been considered

I can see why Feeds makes these required, but the fact it provides an explanation that is not visible anywhere means it is confusing for users.

## Additional information

```
a:13:{s:4:"name";s:21:"Feeds JSONPath Parser";s:11:"description";s:37:"Parse a JSON document using JSONPath.";s:7:"package";s:5:"Feeds";s:8:"backdrop";s:3:"1.x";s:4:"type";s:6:"module";s:12:"dependencies";a:2:{i:0;s:5:"feeds";i:1;s:9:"xautoload";}s:5:"mtime";i:1689679314;s:7:"version";N;s:3:"php";s:5:"5.6.0";s:5:"files";a:0:{}s:9:"bootstrap";i:0;s:8:"required";b:1;s:11:"explanation";s:99:"Feeds is currently using this module for one or more <a href="/admin/structure/feeds">importers</a>";}
```

Add any other information or screenshots that would help.

![image](https://github.com/backdrop/backdrop-issues/assets/20357797/d47f07a2-d1b3-4800-a33d-5eca1dda62e7)

![image](https://github.com/backdrop/backdrop-issues/assets/20357797/48fa6722-cc49-4988-b88e-bb37454df714)

## Draft of feature description for Press Release (1 paragraph at most)

Backdrop now displays required contributed modules in the list of modules and provides the explanation, if given, for why it is required.

**Comment by stpaultim** [comment]
Never run into this before. It certainly sounds like it could be very confusing and that should be handled better. Was this a sub-module of another module?

**Comment by yorkshire-pudding** [comment]
No, `feeds_jsonpath_parser` is [a module](https://github.com/backdrop-contrib/feeds_jsonpath_parser); it is a plugin of feeds, but this will happen for any plugin for feeds that is in use within an importer.

**Comment by indigoxela** [comment]
Still struggling to reproduce...

What I tried so far:

- Install and enable feeds (including dependencies), feeds UI, and feeds_jsonpath_parser
- Create an importer using JSONPath parser, and a dummy mapping

The module still shows up in the listing and isn't even disabled to uninstall. Which step am I missing?

Note: testing happens on PHP 8.1 and it nags "Warning: Undefined array key "module" in feeds_system_info_alter() (line 896 of ...modules/feeds/feeds.module).

**Comment by indigoxela** [comment]
Confirmed, it's very likely a core bug.

The problem arises, when setting "required_by_distribution" even for contrib modules in function system_modules():

https://github.com/backdrop/backdrop/blob/1.x/core/modules/system/system.admin.inc#L616

Disabling is correct, hiding isn't. Modules that are "required_by_distribution" get hidden.

**Comment by kiamlaluno** [comment]
A `hook_system_info_alter()` implementation should not mark as required a module that is disabled.

**Comment by indigoxela** [comment]
> A hook_system_info_alter() implementation should not mark as required a module that is disabled.

There seem to be additional issues with the module, but as I reproduced the reported problem (see my comment), the core behavior is totally odd. It marks the enabled contrib module as "required_by_distribution", just because it is required. That doesn't make much sense to me. :wink:

**Comment by argiepiano** [comment]
I ran into this issue with a different Feeds submodule. Confirming that, once you create a feeds importer that uses the plugins defined by the module, `feeds_system_info_alter()` marks the module as `required`. 

Backdrop core hides the modules from the list when they are `required`. I don't consider this a bug. There is another option which will disable its checkbox: `disabled`. This option is not documented in the [Creating modules page](https://docs.backdropcms.org/creating-modules). In fact some core modules use this option - see `field_system_info_alter()` (found [here](https://github.com/backdrop/backdrop/blob/909f50d6b047f52ca65d2ecef0cab2fbecd74a07/core/modules/field/field.module#L331)).

So, my suggestion is, rather than "fixing" this in core (which, in my opinion, works as expected), we should fix this in Feeds (see my [issue](https://github.com/backdrop-contrib/feeds/issues/141) and PR). And, we should document the option to mark modules' checkboxes as disabled.

As a side note: the choice of the word `disabled` is unfortunate, as it may confuse people into thinking that setting `$info['disabled'] = TRUE` will disable the module. It doesn't. It just disables the checkbox.

**Comment by yorkshire-pudding** [comment]
> As a side note: the choice of the word `disabled` is unfortunate, as it may confuse people into thinking that setting `$info['disabled'] = TRUE` will disable the module. It doesn't. It just disables the checkbox.

I wonder if 'locked' would be a more accurate word to use and I wonder how widespread it is and how difficult it would be to change.

**Comment by kiamlaluno** [comment]
The issue in this case is that a module is listed under _Required by_, as _enabled_, but then it is not present in the module list. It is quite confusing.
I am not sure that Backdrop core can assume no module can mark another module as required.

**Comment by argiepiano** [comment]
> I wonder if 'locked' would be a more accurate word to use and I wonder how widespread it is and how difficult it would be to change.

I like that (`locked` instead of `disabled` to indicate the checkbox is disabled). I did a search in contrib. The only  module that uses `$info['disabled']` is mailsystem.

**Comment by klonos** [comment]
> I like that (`locked` instead of `disabled` to indicate the checkbox is disabled). ...

Me too! 👍🏼 

> ... The only  module that uses `$info['disabled']` is mailsystem. 

If we kept `disabled` as a backwards compatible option but discouraged it (with a deprecated warning), and replace the that with `locked`, then I think that this would work nicely.

**Comment by NormPlum** [comment]
Does this issue need to remain open? It looks like Feeds was fixed, based on @argiepiano's PR and comments that core works as designed.

**Comment by yorkshire-pudding** [comment]
@NormPlum - thanks for pointing this out.  I have created a new issue to capture the aspiration of changing the 'disabled' attribute to 'locked' at #6998
